### ISSUE TITLE
Bad scaling behavior and excessive memory consumption for certain type of dask graphs

### ISSUE THREAD

**Issue by magonser** [dask/distributed#1203]
Hello,

I've been posting a message on this issue before on [stackoverflow](https://stackoverflow.com/questions/44083819/dask-example-with-excessive-memory-consumption-in-distributed-scheduler/). The user helping me suggested, that the `distributed`scheduler only scales well for emberassingly parallel cases, which this example might not be. Since that post, I've tried to further optimise my approaches, but have not been satisfied with my results. Let me explain:

The basic graph structure of question is, where multiple one dimensional arrays are broadcast against each other and subsequently some user defined function is applied to the "meshgrid", before results might or might not get reduced in dimensionality again. The graph therefore has a wide middle section and is narrow in the beginning and the end. Because `numpy` usually creates a view for the `broadcast_to` operation, the native `dask` threaded schedulers compute the aforementioned graph structure just fine. However `distributed` scheduler seems to create copies of the "meshgrid" and hence create lot's of intermediate data. If `bcolz` or `lz4` is installed, the problem is somewhat diminshed, but overall the performance is still really bad. 

Additionally there seems to be the added problem, that so far I have not been able to process a graph with `distributed` scheduler, where aforementioned wide middle section contains a total amount of arrays equivalent to more than the RAM of the computer. Watching memory consumption, all RAM will be consumed completely and `distributed` starts utilising write-to-disk routines internally. Again the `dask` threaded scheduler calculates the graph without problems and with small RAM footprint. Please note, that I set the chunk size small enough.

By usage of `dask.optimize` routines, such as `dask.optimize.fuse` I could not solve the problem. The only thing which helped, was to rebuild the dask graph to different "architecture" by duplicating tasks in the graph, so the graphs structure changes from a diamond structure to a pyramidal structure. Of course with the drawback, that many tasks are calculated several times. This is also partially unsatisfactory solution to me, because my real examples are with hugh intermediate parameter space and amount of dimensions, and the described graph rebuild in itself (from diamond to pyramid) becomes massive .

Further, after reading Issue  #1155, I was expecting the by the use of blosc, `distributed` scheduler would finally be able to perform similar to native `dask` scheduler. I believed so, because in mentioned Issue it was stated, that `blosc` can handle numpy views. Unfortunately the excessive RAM consumption still takes place.

Attached is an [ipython notebook](https://github.com/dask/distributed/files/1100345/dask_distributed_excessive_RAM_usage.ipynb.zip), which can show you the problem. Please create an environment with 

    $conda create -n test python=3.6 jupyter numpy dask distributed blosc
    $source activate test
    $jupyter notebook dask_distributed_excessive_RAM_usage.ipynb

I appreciate `dask` and `distributed` library a lot - it is awesome work. Hopefully I've been able to convey my point, that the described problem (too big intermediate data size, but by chunking still small to fit into the RAM if scheduled correctly) is an example for which, in my opinion, conceptually dask/distributed should be well suited for, but in reality does underperform. 

I'd appreciate a look at the issue and give me hints if I missed out on something, or hint on what could be done within `distributed` to make performance better here. 

Markus

[dask_distributed_excessive_RAM_usage.ipynb.zip](https://github.com/dask/distributed/files/1100345/dask_distributed_excessive_RAM_usage.ipynb.zip)

**Comment by mrocklin** [comment]
I've posted an answer on stack overflow.  I'm not sure how helpful it will be though.  Generally when trying to identify scheduler flaws the first thing to do is to reduce the problem as much as possible while still maintaining the central flaw.

**Comment by magonser** [comment]
I am trying to further nail down the issue. It will take me a couple more days. Where shall I post the reduced example once it is ready? Here or on stackoverflow?

**Comment by mrocklin** [comment]
Generally if it's a usage question then stack overflow, if you think it's a bug then github.  If you're not sure then either will be fine.

**Comment by mrocklin** [comment]
Any update here?  Can this be closed?

**Comment by magonser** [comment]
Well, I am not done with the further dissection of the example. I am stil on it. I appreciate your recent [blog](http://matthewrocklin.com/blog/work/2017/07/03/scaling)-entry very much! My goal is to further provide insight in memory-side of the scaling behaviour for the graph patterns. If progress on my side is too slow, please close or postpone the issue. We can reopen it or discuss on [stackoverflow](https://stackoverflow.com/questions/44083819/dask-example-with-excessive-memory-consumption-in-distributed-scheduler/)

**Comment by magonser** [comment]
Hello ,

so here is the simplified example for my follow up on [stackoverflow question](https://stackoverflow.com/questions/44083819/dask-example-with-excessive-memory-consumption-in-distributed-scheduler/).

[dask_distributed_RAM_usage_meshgrid_operation.ipynb.zip](https://github.com/dask/distributed/files/1168096/dask_distributed_RAM_usage_meshgrid_operation.ipynb.zip)

**Comment by magonser** [comment]
Closing, because it is outdated
